File: CS/Pivot-Aggr-Examples/CodeFiles/CodeExample.cs
using DevExpress.Data.PivotGrid;
using DevExpress.XtraPivotGrid;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Pivot_Aggr_Examples
{
    static class CodeExample
    {
        public static PivotGridControl pivotGrid { get; set; }
        public static void SalesByYearByCountryExample()
        {
            #region Filter Sales by Year and Country
            PivotGridField salesByYearByCountry = new PivotGridField()
            {
                Name = "fldSalesByYearByCountry",
                Area = PivotArea.FilterArea,
                Caption = "Sales By Year By Country",
                Visible = false,
                FieldName = "SalesByYearByCountry"
            };
            salesByYearByCountry.DataBinding = new ExpressionDataBinding("Aggr(Sum([Extended Price]), GetYear([OrderDate]), [Country])");
            pivotGrid.Fields.Add(salesByYearByCountry);
            pivotGrid.ActiveFilter.CriteriaString = "[SalesByYearByCountry] > 300000";
            #endregion
            pivotGrid.BestFit();
        }

        public static void SalesByYearByCountryExample_Clear()
        {
            pivotGrid.ActiveFilter.Clear();
            PivotGridField fld = pivotGrid.Fields.GetFieldByName("fldSalesByYearByCountry");
            if (fld != null)
                pivotGrid.Fields.Remove(fld);                
        }

        public static void AverageSalesByYearExample()
        {
            #region Average Sales by Year
            PivotGridField salesByYearByCountry = new PivotGridField() {
                Name = "fldSalesByYearByCountry",
                Area = PivotArea.DataArea,
                Caption = "Average Sales by Year",
                SummaryType = PivotSummaryType.Average
            };
            salesByYearByCountry.CellFormat.FormatType = DevExpress.Utils.FormatType.Custom;
            salesByYearByCountry.CellFormat.FormatString = "c";
            salesByYearByCountry.DataBinding = new ExpressionDataBinding("Aggr(Sum([Extended Price]), GetYear([OrderDate]), [Country])");
            salesByYearByCountry.Options.ShowValues = false;
            pivotGrid.Fields.Add(salesByYearByCountry);
            #endregion
            pivotGrid.BestFit();
        }

        public static void AverageSalesByYearExample_Clear()
        {
            PivotGridField fld = pivotGrid.Fields.GetFieldByName("fldSalesByYearByCountry");
            if (fld != null)
                pivotGrid.Fields.Remove(fld);
        }

        public static void YearOfFirstSaleExample()
        {
            #region Grouping by Year of First Sale
            PivotGridField yearOfFirstSale = new PivotGridField() {
                Name = "fldYearOfFirstSale",
                Area = PivotArea.RowArea,
                AreaIndex = 0,
                Caption = "Year of first sale" };
            yearOfFirstSale.DataBinding = new ExpressionDataBinding("Aggr(Min(GetYear([OrderDate])), [Country])");
            yearOfFirstSale.Options.ShowTotals = false;
            pivotGrid.Fields.Add(yearOfFirstSale);
            #endregion
            pivotGrid.BestFit();
        }

        public static void YearOfFirstSaleExample_Clear()
        {
            PivotGridField fld = pivotGrid.Fields.GetFieldByName("fldYearOfFirstSale");
            if (fld != null)
                pivotGrid.Fields.Remove(fld);
        }

        public static void PercentOfTotalExample()
        {
            #region Percent of Total
            // Calculates the percentage of the grand total column. 
            // You can also use the PivotGridField.SummaryDisplayType property to display summaries, 
            // but the Aggr expression allows you to sort by the summary column and use the values in subsequent calculations.

            PivotGridField percentOfTotal = new PivotGridField()
            {
                Name = "fldPercentOfTotal",
                Area = PivotArea.DataArea,
                AreaIndex = 0,
                Caption = "% of Total"
            };
            percentOfTotal.DataBinding = new ExpressionDataBinding("Sum([Extended Price]) / Min(Aggr(Sum([Extended Price]), [Country]))");
            percentOfTotal.CellFormat.FormatType = DevExpress.Utils.FormatType.Custom;
            percentOfTotal.CellFormat.FormatString = "P";
            percentOfTotal.Options.ShowTotals = false;
            percentOfTotal.Options.ShowGrandTotal = false;
            pivotGrid.Fields.Add(percentOfTotal);
            pivotGrid.OptionsView.ShowRowGrandTotals = false;
            #endregion
            pivotGrid.BestFit();
        }

        public static void PercentOfTotalExample_Clear()
        {
            PivotGridField fld = pivotGrid.Fields.GetFieldByName("fldPercentOfTotal");
            if (fld != null)
                pivotGrid.Fields.Remove(fld);
        }

        public static void ShowTotalsInRowAreaExample()
        {

            PivotGridField fieldOrderMonth1 = new PivotGridField
            {
                Area = PivotArea.ColumnArea,
                AreaIndex = 1,
                Caption = "Month",
                FieldName = "OrderDate",
                GroupInterval = PivotGroupInterval.DateMonth,
                Name = "fieldOrderMonth1",
            };
            pivotGrid.Fields.Add(fieldOrderMonth1);

            pivotGrid.Fields.GetFieldByName("fieldCountry1").Area = PivotArea.FilterArea;
            pivotGrid.Fields.GetFieldByName("fieldCategoryName1").Area = PivotArea.RowArea;
            #region Show Totals in Row Area
            // Displays total values in the Row Area, as a "standalone column".

            PivotGridField amountMax = new PivotGridField()
            {
                Name = "fldAmountMax",
                Area = PivotArea.RowArea,
                AreaIndex = 1,
                Caption = "Max Amount per Month"
            };
            amountMax.DataBinding = new ExpressionDataBinding("Aggr(Max(Aggr(Sum([Extended Price]), [fieldCategoryName1], [fieldOrderDate1], [fieldOrderMonth1])), [fieldCategoryName1])");
            amountMax.ValueFormat.FormatType = DevExpress.Utils.FormatType.Custom;
            amountMax.ValueFormat.FormatString = "C2";
            pivotGrid.Fields.Add(amountMax);
            #endregion
            pivotGrid.BestFit();
        }

        public static void ShowTotalsInRowAreaExample_Clear()
        {
            PivotGridField fld = pivotGrid.Fields.GetFieldByName("fldAmountMax");
            if (fld != null)
                pivotGrid.Fields.Remove(fld);
            PivotGridField fld1 = pivotGrid.Fields.GetFieldByName("fieldOrderMonth1");
            if (fld1 != null)
                pivotGrid.Fields.Remove(fld1);
            pivotGrid.Fields.GetFieldByName("fieldCountry1").Area = PivotArea.RowArea;
            pivotGrid.Fields.GetFieldByName("fieldCategoryName1").Area = PivotArea.FilterArea;
        }
    }
}

File: VB/Pivot-Aggr-Examples/CodeFiles/CodeExample.vb
Imports DevExpress.Data.PivotGrid
Imports DevExpress.XtraPivotGrid
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks

Namespace Pivot_Aggr_Examples
    Friend NotInheritable Class CodeExample

        Private Sub New()
        End Sub

        Public Shared Property pivotGrid() As PivotGridControl
        Public Shared Sub SalesByYearByCountryExample()
            '			#Region "Filter Sales by Year and Country"
            Dim salesByYearByCountry As New PivotGridField()
            salesByYearByCountry.Name = "fldSalesByYearByCountry"
            salesByYearByCountry.Area = PivotArea.FilterArea
            salesByYearByCountry.Caption = "Sales By Year By Country"
            salesByYearByCountry.Visible = False
            salesByYearByCountry.FieldName = "SalesByYearByCountry"
            salesByYearByCountry.DataBinding = New ExpressionDataBinding("Aggr(Sum([Extended Price]), GetYear([OrderDate]), [Country])")
            pivotGrid.Fields.Add(salesByYearByCountry)
            pivotGrid.ActiveFilter.CriteriaString = "[SalesByYearByCountry] > 300000"
            '			#End Region
            pivotGrid.BestFit()
        End Sub

        Public Shared Sub SalesByYearByCountryExample_Clear()
            pivotGrid.ActiveFilter.Clear()
            Dim fld As PivotGridField = pivotGrid.Fields.GetFieldByName("fldSalesByYearByCountry")
            If fld IsNot Nothing Then
                pivotGrid.Fields.Remove(fld)
            End If
        End Sub

        Public Shared Sub AverageSalesByYearExample()
            '			#Region "Average Sales by Year"
            Dim salesByYearByCountry As New PivotGridField()
            salesByYearByCountry.Name = "fldSalesByYearByCountry"
            salesByYearByCountry.Area = PivotArea.DataArea
            salesByYearByCountry.Caption = "Average Sales by Year"
            salesByYearByCountry.SummaryType = PivotSummaryType.Average
            salesByYearByCountry.CellFormat.FormatType = DevExpress.Utils.FormatType.Custom
            salesByYearByCountry.CellFormat.FormatString = "c"
            salesByYearByCountry.DataBinding = New ExpressionDataBinding("Aggr(Sum([Extended Price]), GetYear([OrderDate]), [Country])")
            salesByYearByCountry.Options.ShowValues = False
            pivotGrid.Fields.Add(salesByYearByCountry)
            '			#End Region
            pivotGrid.BestFit()
        End Sub

        Public Shared Sub AverageSalesByYearExample_Clear()
            Dim fld As PivotGridField = pivotGrid.Fields.GetFieldByName("fldSalesByYearByCountry")
            If fld IsNot Nothing Then
                pivotGrid.Fields.Remove(fld)
            End If
        End Sub

        Public Shared Sub YearOfFirstSaleExample()
            '			#Region "Grouping by Year of First Sale"
            Dim yearOfFirstSale As New PivotGridField()
            yearOfFirstSale.Name = "fldYearOfFirstSale"
            yearOfFirstSale.Area = PivotArea.RowArea
            yearOfFirstSale.AreaIndex = 0
            yearOfFirstSale.Caption = "Year of first sale"
            yearOfFirstSale.DataBinding = New ExpressionDataBinding("Aggr(Min(GetYear([OrderDate])), [Country])")
            yearOfFirstSale.Options.ShowTotals = False
            pivotGrid.Fields.Add(yearOfFirstSale)
            '			#End Region
            pivotGrid.BestFit()
        End Sub

        Public Shared Sub YearOfFirstSaleExample_Clear()
            Dim fld As PivotGridField = pivotGrid.Fields.GetFieldByName("fldYearOfFirstSale")
            If fld IsNot Nothing Then
                pivotGrid.Fields.Remove(fld)
            End If
        End Sub

        Public Shared Sub PercentOfTotalExample()
            '			#Region "Percent of Total"
            ' Calculates the percentage of the grand total column. 
            ' You can also use the PivotGridField.SummaryDisplayType property to display summaries, 
            ' but the Aggr expression allows you to sort by the summary column and use the values in subsequent calculations.

            Dim percentOfTotal As New PivotGridField()
            percentOfTotal.Name = "fldPercentOfTotal"
            percentOfTotal.Area = PivotArea.DataArea
            percentOfTotal.AreaIndex = 0
            percentOfTotal.Caption = "% of Total"
            percentOfTotal.DataBinding = New ExpressionDataBinding("Sum([Extended Price]) / Min(Aggr(Sum([Extended Price]), [Country]))")
            percentOfTotal.CellFormat.FormatType = DevExpress.Utils.FormatType.Custom
            percentOfTotal.CellFormat.FormatString = "P"
            percentOfTotal.Options.ShowTotals = False
            percentOfTotal.Options.ShowGrandTotal = False
            pivotGrid.Fields.Add(percentOfTotal)
            pivotGrid.OptionsView.ShowRowGrandTotals = False
            '			#End Region
            pivotGrid.BestFit()
        End Sub

        Public Shared Sub PercentOfTotalExample_Clear()
            Dim fld As PivotGridField = pivotGrid.Fields.GetFieldByName("fldPercentOfTotal")
            If fld IsNot Nothing Then
                pivotGrid.Fields.Remove(fld)
            End If
        End Sub

        Public Shared Sub ShowTotalsInRowAreaExample()

            Dim fieldOrderMonth1 As PivotGridField = New PivotGridField()
            fieldOrderMonth1.Area = PivotArea.ColumnArea
            fieldOrderMonth1.AreaIndex = 1
            fieldOrderMonth1.Caption = "Month"
            fieldOrderMonth1.FieldName = "OrderDate"
            fieldOrderMonth1.GroupInterval = PivotGroupInterval.DateMonth
            fieldOrderMonth1.Name = "fieldOrderMonth1"
            pivotGrid.Fields.Add(fieldOrderMonth1)

            pivotGrid.Fields.GetFieldByName("fieldCountry1").Area = PivotArea.FilterArea
            pivotGrid.Fields.GetFieldByName("fieldCategoryName1").Area = PivotArea.RowArea
            '			#Region "Show Totals in Row Area"
            ' Displays total values in the Row Area, as a "standalone column".

            Dim amountMax As New PivotGridField()
            amountMax.Name = "fldAmountMax"
            amountMax.Area = PivotArea.RowArea
            amountMax.AreaIndex = 1
            amountMax.Caption = "Max Amount per Month"
            amountMax.DataBinding = New ExpressionDataBinding("Aggr(Max(Aggr(Sum([Extended Price]), [fieldCategoryName1], [fieldOrderDate1], [fieldOrderMonth1])), [fieldCategoryName1])")
            amountMax.ValueFormat.FormatType = DevExpress.Utils.FormatType.Custom
            amountMax.ValueFormat.FormatString = "C2"
            pivotGrid.Fields.Add(amountMax)
            '			#End Region
            pivotGrid.BestFit()
        End Sub

        Public Shared Sub ShowTotalsInRowAreaExample_Clear()
            Dim fld As PivotGridField = pivotGrid.Fields.GetFieldByName("fldAmountMax")
            If fld IsNot Nothing Then
                pivotGrid.Fields.Remove(fld)
            End If
            Dim fld1 As PivotGridField = pivotGrid.Fields.GetFieldByName("fieldOrderMonth1")
            If fld1 IsNot Nothing Then
                pivotGrid.Fields.Remove(fld1)
            End If
            pivotGrid.Fields.GetFieldByName("fieldCountry1").Area = PivotArea.RowArea
            pivotGrid.Fields.GetFieldByName("fieldCategoryName1").Area = PivotArea.FilterArea
        End Sub
    End Class
End Namespace
